pkg/blobinfocache/internal/prioritize/prioritize.go

```diff
@@ -77,34 +82,75 @@ func (css *candidateSortState) Swap(i, j int) {
 	css.cs[i], css.cs[j] = css.cs[j], css.cs[i]
 }
 
-// destructivelyPrioritizeReplacementCandidatesWithMax is destructivelyPrioritizeReplacementCandidates with a parameter for the
-// number of entries to limit, only to make testing simpler.
-func destructivelyPrioritizeReplacementCandidatesWithMax(cs []CandidateWithTime, primaryDigest, uncompressedDigest digest.Digest, maxCandidates int) []blobinfocache.BICReplacementCandidate2 {
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// destructivelyPrioritizeReplacementCandidatesWithMax is destructivelyPrioritizeReplacementCandidates with parameters for the
+// number of entries to limit for known and unknown location separately, only to make testing simpler.
+// TODO: following function is not destructive any more in the nature instead priortized result is actually copies of the original
+// candidate set, so In future we might wanna re-name this public API and remove the destructive prefix.
+func destructivelyPrioritizeReplacementCandidatesWithMax(cs []CandidateWithTime, primaryDigest, uncompressedDigest digest.Digest, maxCandidatesLimit int, unknownLocationLimit int) []blobinfocache.BICReplacementCandidate2 {
+	// split unknown candidates and known candidates
+	// and limit them seperately.
+	var knownLocationCandidates []CandidateWithTime
+	var unknownLocationCandidates []CandidateWithTime
+
+	for _, candidate := range cs {
+		if candidate.Candidate.UnknownLocation {
+			unknownLocationCandidates = append(unknownLocationCandidates, candidate)
+		} else {
+			knownLocationCandidates = append(knownLocationCandidates, candidate)
+		}
+	}
+
 	// We don't need to use sort.Stable() because nanosecond timestamps are (presumably?) unique, so no two elements should
 	// compare equal.
 	sort.Sort(&candidateSortState{
-		cs:                 cs,
+		cs:                 knownLocationCandidates,
+		primaryDigest:      primaryDigest,
+		uncompressedDigest: uncompressedDigest,
+	})
+	sort.Sort(&candidateSortState{
+		cs:                 unknownLocationCandidates,
 		primaryDigest:      primaryDigest,
 		uncompressedDigest: uncompressedDigest,
 	})
 
-	resLength := len(cs)
-	if resLength > maxCandidates {
-		resLength = maxCandidates
+	knownUsed := min(len(knownLocationCandidates), maxCandidatesLimit)
+	unknownUsed := min(len(unknownLocationCandidates), min(unknownLocationLimit, knownUsed))
```

Comment:
This looks incorrect. If `knownUsed == 0`, `unknownUsed` will also be 0.

Compare https://github.com/containers/image/pull/1645#discussion_r970047362 .

Introduce a
```go
remainingCapacity := maxCandidates-knownUsed
```
if it helps.